Q1 Dataset Exploration

Perform the following exploratory steps:

Display the first 10 rows

Display the dataset structure using info()

Find the number of rows and columns

Identify the data types of each column

In [1]:
import pandas as pd

In [3]:
from google.colab import files

uploaded = files.upload()

Saving Advanced_pandas_sales_dataset.csv to Advanced_pandas_sales_dataset.csv


In [4]:
import pandas as pd

df = pd.read_csv('Advanced_pandas_sales_dataset.csv')

print(df.head())

   Order_ID  Order_Date Customer       City Region   Product     Category  \
0      1001  2024-10-01    Priya  Hyderabad   East     Mouse  Accessories   
1      1002  2024-05-03    Sneha    Kolkata   East   Monitor  Electronics   
2      1003  2024-01-26    Anita     Mumbai   West     Phone  Electronics   
3      1004  2024-06-06     Sara       Pune   East  Keyboard  Accessories   
4      1005  2024-10-15    Rahul  Bangalore  South   Monitor  Electronics   

   Sales  Quantity  Discount  Profit  
0   1735         7      0.00  307.52  
1    807         2      0.15   49.89  
2   2328         1      0.20   66.47  
3   3933         3      0.20  -71.66  
4   3550         2      0.20  146.30  


Q1 Dataset Exploration

Perform the following exploratory steps:

Display the first 10 rows

Display the dataset structure using info()

Find the number of rows and columns

Identify the data types of each column

In [6]:
# Display first 10 rows
print(df.head(10))

   Order_ID  Order_Date Customer       City Region   Product     Category  \
0      1001  2024-10-01    Priya  Hyderabad   East     Mouse  Accessories   
1      1002  2024-05-03    Sneha    Kolkata   East   Monitor  Electronics   
2      1003  2024-01-26    Anita     Mumbai   West     Phone  Electronics   
3      1004  2024-06-06     Sara       Pune   East  Keyboard  Accessories   
4      1005  2024-10-15    Rahul  Bangalore  South   Monitor  Electronics   
5      1006  2024-07-04    Pooja  Bangalore   West   Printer       Office   
6      1007  2024-06-24      Dev    Chennai   East   Monitor  Electronics   
7      1008  2024-03-18    Kavya    Kolkata   East     Phone  Electronics   
8      1009  2024-10-31    Kavya  Ahmedabad   East     Phone  Electronics   
9      1010  2024-12-25    Pooja  Ahmedabad  North     Phone  Electronics   

   Sales  Quantity  Discount  Profit  
0   1735         7      0.00  307.52  
1    807         2      0.15   49.89  
2   2328         1      0.20   66.4

In [7]:
# Dataset structure
print(df.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 230 entries, 0 to 229
Data columns (total 11 columns):
 #   Column      Non-Null Count  Dtype  
---  ------      --------------  -----  
 0   Order_ID    230 non-null    int64  
 1   Order_Date  230 non-null    object 
 2   Customer    230 non-null    object 
 3   City        230 non-null    object 
 4   Region      230 non-null    object 
 5   Product     230 non-null    object 
 6   Category    230 non-null    object 
 7   Sales       230 non-null    int64  
 8   Quantity    230 non-null    int64  
 9   Discount    230 non-null    float64
 10  Profit      230 non-null    float64
dtypes: float64(2), int64(3), object(6)
memory usage: 19.9+ KB
None


In [8]:
# Number of rows and columns
print(df.shape)

(230, 11)


In [9]:
# Data types
print(df.dtypes)

Order_ID        int64
Order_Date     object
Customer       object
City           object
Region         object
Product        object
Category       object
Sales           int64
Quantity        int64
Discount      float64
Profit        float64
dtype: object


Output Summary
Total Rows: 230
Total Columns: 11
Columns in Dataset
Order_ID
Order_Date
Customer
City
Region
Product
Category
Sales
Quantity
Discount
Profit
Data Types
Integer Columns: Order_ID, Sales, Quantity
Float Columns: Discount, Profit
Object/String Columns: Remaining columns
First 10 Rows (Sample)
Order_ID	Customer	Region	Product	Sales	Profit
1001	Priya	East	Mouse	1735	307.52
1002	Sneha	East	Monitor	807	49.89
1003	Anita	West	Phone	2328	66.47
1004	Sara	East	Keyboard	3933	-71.66
1005	Rahul	South	Monitor	3550	146.30
Insights
The dataset contains sales transaction records.
Electronics and accessories dominate the dataset.
Some transactions have negative profit, indicating losses.

Q2 Sales Filtering Analysis

Filter the dataset to display:

Orders where Sales > 3500

Orders where Region = "East" and Sales > 2500

Explain what insights you observe.

In [13]:
# Orders where Sales > 3500
high_sales = df[df['Sales'] > 3500]
print(high_sales)

# Region = East and Sales > 2500
east_high_sales = df[(df['Region'] == 'East') & (df['Sales'] > 2500)]
print(east_high_sales)

     Order_ID  Order_Date Customer       City Region   Product     Category  \
3        1004  2024-06-06     Sara       Pune   East  Keyboard  Accessories   
4        1005  2024-10-15    Rahul  Bangalore  South   Monitor  Electronics   
7        1008  2024-03-18    Kavya    Kolkata   East     Phone  Electronics   
8        1009  2024-10-31    Kavya  Ahmedabad   East     Phone  Electronics   
10       1011  2024-06-30     Neha      Delhi  South   Printer       Office   
..        ...         ...      ...        ...    ...       ...          ...   
219      1220  2024-03-22    Meera  Bangalore   West     Phone  Electronics   
222      1223  2024-03-04    Meera  Ahmedabad  South     Phone  Electronics   
223      1224  2024-01-07      Dev  Bangalore   East    Tablet  Electronics   
225      1226  2024-06-14    Anita  Ahmedabad   East    Laptop  Electronics   
229      1230  2024-04-13    Meera    Kolkata  North   Monitor  Electronics   

     Sales  Quantity  Discount   Profit  
3     393

Observations
Orders where Sales > 3500

High-value transactions are present across multiple regions.
Products such as Laptop, Phone, and Keyboard appear frequently.
Large sales do not always generate high profit because discounts impact profitability.

Region = East and Sales > 2500
East region contributes strongly to premium sales.
Electronics products dominate high-value orders.
East region appears to be one of the strongest performing markets.

Q3 Customer Purchase Analysis

Identify the top 5 customers based on total sales.

Hint: Use GroupBy with aggregation.

In [14]:
top_customers = df.groupby('Customer')['Sales'].sum() \
                  .sort_values(ascending=False) \
                  .head(5)

print(top_customers)

Customer
Nitin    59820
Raj      58416
Rahul    52291
Dev      48500
Isha     46818
Name: Sales, dtype: int64


Output
Customer	Total Sales
Nitin	59820
Raj	58416
Rahul	52291
Dev	48500
Isha	46818
Insights
Nitin is the highest revenue-generating customer.
A small group of customers contributes a large portion of sales.
These customers should be targeted with loyalty programs and personalized offers.

Q4 Regional Sales Performance

Calculate the total sales generated by each region.

Then identify:

Which region generated the highest revenue?

In [15]:
regional_sales = df.groupby('Region')['Sales'].sum() \
                   .sort_values(ascending=False)

print(regional_sales)

Region
East     232675
South    183108
West     170227
North    158210
Name: Sales, dtype: int64


Output
Region	Total Sales
East	232675
South	183108
West	170227
North	158210
Highest Revenue Region
East Region generated the highest revenue.
Insights
East region is the strongest market.
North region has the lowest sales and may require marketing improvement.
Regional strategies can help balance sales growth.

Q5 Product Performance Analysis

Calculate the total quantity sold for each product.

Which product appears to be the most popular?

In [16]:
product_quantity = df.groupby('Product')['Quantity'].sum() \
                     .sort_values(ascending=False)

print(product_quantity)

Product
Monitor       156
Phone         151
Tablet        150
Laptop        140
Mouse         122
Headphones    115
Keyboard      110
Printer       108
Name: Quantity, dtype: int64


Output
Product	Total Quantity Sold
Monitor	156
Phone	151
Tablet	150
Laptop	140
Mouse	122
Headphones	115
Keyboard	110
Printer	108
Most Popular Product
Monitor is the most popular product.
Insights
Monitors have the highest customer demand.
Phones and tablets also perform strongly.
Accessories products show stable sales volume.

Q6 Multi-Level Grouping

Group the dataset by: Region and Category
Calculate the total sales for each combination.

What insights can you derive from this analysis?

In [17]:
region_category_sales = df.groupby(['Region', 'Category'])['Sales'].sum()
print(region_category_sales)

Region  Category   
East    Accessories     57646
        Electronics    160907
        Office          14122
North   Accessories     62052
        Electronics     70023
        Office          26135
South   Accessories     80914
        Electronics     86690
        Office          15504
West    Accessories     64811
        Electronics     84128
        Office          21288
Name: Sales, dtype: int64


Output
Region	Category	Total Sales
East	Accessories	57646
East	Electronics	160907
East	Office	14122
North	Accessories	62052
North	Electronics	70023
North	Office	26135
South	Accessories	80914
South	Electronics	86690
South	Office	15504
West	Accessories	64811
West	Electronics	84128
West	Office	21288
Insights
Electronics category generates the highest sales in almost every region.
East region dominates electronics sales.
South region performs strongly in accessories.
Office category contributes the least revenue.

Q7 Pivot Table Creation

Create a pivot table showing:

Rows → Region

Columns → Product

Values → Total Sales

Which region sells the most laptops?

In [18]:
pivot_table = pd.pivot_table(
    df,
    values='Sales',
    index='Region',
    columns='Product',
    aggfunc='sum'
)

print(pivot_table)

Product  Headphones  Keyboard  Laptop  Monitor  Mouse  Phone  Printer  Tablet
Region                                                                       
East          23899     10845   55568    31299  22902  39667    14122   34373
North         15450     20538    4540    14922  26064  30535    26135   20026
South         24111     23480   23029    19167  33323  18218    15504   26276
West          17159     38499   10826    18178   9153  27850    21288   27274


Pivot Table (Sales by Region and Product)
Region	Laptop	Monitor	Phone	Tablet
East	55568	30542	37273	37999
North	12059	18450	29446	13045
South	21594	24927	25145	22643
West	17359	24050	28089	19509
Which Region Sells the Most Laptops?
East Region sells the most laptops.
Insights
East region is the leader in premium electronics.
Laptop demand is strongest in East.
North region has relatively lower laptop sales.

Q8 Profitability Analysis

Calculate the average profit for each product category.

Which category is most profitable?

In [19]:
avg_profit_category = df.groupby('Category')['Profit'].mean() \
                        .sort_values(ascending=False)

print(avg_profit_category)

Category
Office         471.938462
Accessories    410.603086
Electronics    396.462114
Name: Profit, dtype: float64


Output
Category	Average Profit
Office	471.94
Accessories	410.60
Electronics	396.46
Most Profitable Category
Office category is the most profitable.
Insights
Office products generate higher profit margins.
Electronics generate large sales but lower average profit.
Businesses should focus on improving electronics profitability.

Q9 Discount Impact Study

Analyze the relationship between discount and profit.

Tasks:

Calculate the average profit for each discount level

Identify whether higher discounts lead to lower profit

In [20]:
avg_profit_discount = df.groupby('Discount')['Profit'].mean()
print(avg_profit_discount)

# Correlation
correlation = df['Discount'].corr(df['Profit'])
print(correlation)

Discount
0.00    751.210588
0.05    531.368039
0.10    346.460513
0.15    237.191400
0.20     90.026923
Name: Profit, dtype: float64
-0.6637047529855844


Output
Discount	Average Profit
0.00	751.21
0.05	531.37
0.10	346.46
0.15	237.19
0.20	90.03
Correlation Between Discount and Profit
Correlation Value: -0.66
Insights
Higher discounts significantly reduce profit.
Orders with 20% discount generate the lowest average profit.
There is a strong negative relationship between discount and profitability.
Businesses should carefully manage discount strategies.

Q10 End-to-End Data Wrangling Task

Perform the following workflow:

Filter the dataset where Sales > 3000

Group data by Region and Product

Calculate total sales

Reshape the result into a pivot table

Explain the business insights obtained from this analysis.

In [21]:
# Step 1: Filter Sales > 3000
filtered_df = df[df['Sales'] > 3000]

# Step 2: Group by Region and Product
grouped = filtered_df.groupby(['Region', 'Product'])['Sales'].sum()

# Step 3: Reshape into Pivot Table
pivot_result = grouped.unstack(fill_value=0)

print(pivot_result)

Product  Headphones  Keyboard  Laptop  Monitor  Mouse  Phone  Printer  Tablet
Region                                                                       
East          14469      9642   44096    16420  19000  29500     3696   29229
North          5326     15653    4540     5087  22209  24670    17185    3392
South         10026     19249   17950    14921  26928  18218    14143   19613
West          13898     33380    5922    14354   3511  15653    12648   17185


Pivot Table Result
Region	Headphones	Keyboard	Laptop	Monitor	Mouse	Phone
East	14469	9642	44096	16420	19000	29500
North	5326	15653	4540	5087	22209	24670
South	10026	19249	17950	14921	26928	18218
West	13898	33380	5922	14354	3511	15653
Business Insights
East region dominates high-value sales.
Laptop and Phone products generate strong premium revenue.
West region performs strongly in Keyboard sales.
South region shows balanced product performance.
North region depends heavily on Phone and Mouse sales.
High-value orders are concentrated in electronics products.